File: Interview-Pratice-FANNG/Recursion/MergeSort.ts
export const numbers = [99,44,6,2,14,3,54,8,8,9,78,7,2];
const mergeSort = (numbers:number[]):number[] => {
    if(numbers.length == 1)  {
        return numbers;
    }

    const indexMiddle = Math.round(numbers.length / 2);
    const leftArray = numbers.slice(0, indexMiddle);
    const rightArray = numbers.splice(indexMiddle,numbers.length);
  
    return merge(mergeSort(leftArray), mergeSort(rightArray));
}

function merge(left:number[],right:number[]):number[] {
    const sortedItems = [];

    while (left.length && right.length) {
        if (left[0] <= right[0]) {
            sortedItems.push(left.shift()!);
        } else {
            sortedItems.push(right.shift()!);
        }
    }
 
    return [...sortedItems, ...right,...left];
}
console.log(mergeSort(numbers))

File: Desgin-Algorithms/DiveAndConquer/mergeSort.py
def merge(arr_left, arr_right):
    if arr_left == []:
        return arr_right
    elif arr_right == []:
        return arr_left
    else: 
        if (arr_left[0] > arr_right[0]):
            return [arr_left[0]] + merge(arr_left[1:],arr_right)
        else:
            return [arr_right[0]] + merge(arr_left, arr_right[1:])
def merge_sort(arr):
    n = len(arr)
    if n <= 1:
        return arr
    arr_left = merge_sort(arr[0:n//2])
    arr_right = merge_sort(arr[n//2:n])
    return merge(arr_left, arr_right)


print(merge_sort([2,3,4,9,1,0,9,8,3]))
